### PR TITLE
Complete prepare_dump API and apply to occs_aminusb + cleanups

### PR DESCRIPTION
**Pull request by tovrstra**
This is another step in #191, adding a small new feature: the conversion of to unrestricted orbitals when restricted orbitals with `occs_aminusb` is not supported by the selected file format.

This feature is implemented at two levels:

- In `iodata.orbitals` module, a new function `convert_to_unrestricted` is added, which provides the basic functionality. It is useful also outside the context of writing files. Splitting this off also facilitates testing. This is a function (and not a method) on purpose: the class itself does not need it to fulfill its main role, so it can be kept separate.
- In the new `iodata.prepare` module, a function `prepare_unrestricted_aminusb` is created, building on the previous, adding all the logic for errors, warnings and geared towards `prepare_dump` functions of the relevant file formats.

Test coverage of the new code should be fairly complete, but I'll check the CI output to be sure.

There are a few related changes (and a few deferred ones, new tasks in #191):

- Optional arguments in `dump_one` and `dump_many` must now be given as keyword arguments for the sake of clarity. The same change for other functions in `iodata.api` will be made in one of the upcoming PRs.
- An assertion in the Molekel `load_one` function has been turned into a warning. (Eventually, we need to replace all asserts in non-test code. This is on the list in #204) A new test for this warning is included.

I will YOLO-merge this on Friday July 5, unless reviewed earlier.

## Summary by Sourcery

This pull request introduces a new feature to convert restricted orbitals to unrestricted orbitals and integrates this functionality into the `prepare_dump` process for various file formats. It also includes enhancements to the API for clarity, a bug fix for spin polarization handling, and comprehensive test coverage for the new functionality.

- **New Features**:
    - Added a new function `convert_to_unrestricted` in the `iodata.orbitals` module to convert restricted orbitals to unrestricted orbitals.
    - Introduced a new `prepare_unrestricted_aminusb` function in the `iodata.prepare` module to handle conversion logic for `prepare_dump` functions.
- **Bug Fixes**:
    - Changed an assertion in the Molekel `load_one` function to a warning to handle inconsistencies in spin polarization.
- **Enhancements**:
    - Updated `dump_one` and `dump_many` functions in `iodata.api` to require optional arguments as keyword arguments for clarity.
    - Modified `prepare_dump` functions in various format modules (Molekel, Molden, WFN, WFX, FCHK, JSON) to use the new `prepare_unrestricted_aminusb` function for handling `occs_aminusb`.
- **Tests**:
    - Added comprehensive test coverage for the new `convert_to_unrestricted` function in `iodata/test/test_orbitals.py`.
    - Included new tests for the `prepare_unrestricted_aminusb` function in `iodata/test/test_prepare.py`.
    - Added a test in `iodata/test/test_molekel.py` to check for warnings when loading files with incorrect spin multiplicity.

### CODE REVIEW

File: iodata/test/test_orbitals.py [hunk 1]
```diff
@@ -518,3 +518,97 @@ def test_generalized_irreps():
...
+    mo2 = convert_to_unrestricted(mo1)
+    assert mo1 is not mo2
+    assert mo2.kind == "unrestricted"
+    assert mo2.norba == 5
+    assert mo2.norbb == 5
+    assert mo2.coeffs is None
+    assert mo2.occs is None
+    assert mo2.coeffs is None
+    assert mo2.energies is None
+    assert mo2.irreps is None
```
[the remark this answers is not in the document]
Comment (tovrstra) [reply iodata/test/test_orbitals.py:545]: is done elsewhere

File: iodata/orbitals.py [hunk 2]
```diff
@@ -381,3 +381,36 @@ def irrepsb(self):
         if self.kind == "restricted":
             return self.irreps
         return self.irreps[self.norba :]
+
+
+def convert_to_unrestricted(mo: MolecularOrbitals) -> MolecularOrbitals:
```
[the remark this answers is not in the document]
Comment (tovrstra) [reply iodata/orbitals.py:386]: When the helper is implemented correctly, it becomes an ugly thing, going against the original idea of keeping things simple.

**Comment by tovrstra** [comment]
The deepsource issues can be ignored:

- Unused arguments are legit in this case for the sake of API consistency. Deepsource doesn't seem to realize. It requires a more global analysis of the code, not just checking patterns, to see that this is ok.
- The complexity of the molekel load_one function was already present before this PR.

**Comment by tovrstra** [comment]
FYI coverage results look ok: https://github.com/theochem/iodata/actions/runs/9695310238/job/26754784217?pr=352#step:6:1013

**Review by PaulWAyers — APPROVED** [review]
Thanks Toon, I really think this improves the usability.

**Comment by tovrstra** [comment]
Thanks for reviewing, Paul. Merging...

**Pull request merged**